samcli/lib/docker/log_streamer.py

```diff
@@ -0,0 +1,96 @@
+"""
+Log streaming utilities when streaming logs from Docker
+"""
+import os
+from typing import Dict
+
+import docker
+from docker.errors import APIError
+
+from samcli.lib.package.stream_cursor_utils import (
+    CursorUpFormatter,
+    CursorDownFormatter,
+    CursorLeftFormatter,
+    ClearLineFormatter,
+)
+from samcli.lib.utils.stream_writer import StreamWriter
+
+
+class LogStreamer:
+    def __init__(self, stream: StreamWriter, error_class: APIError.__class__, error_msg_prefix=""):
+        self._stream = stream
+        self._error_class = error_class
+        self._error_msg_prefix = error_msg_prefix
+        self._cursor_up_formatter = CursorUpFormatter()
+        self._cursor_down_formatter = CursorDownFormatter()
+        self._cursor_left_formatter = CursorLeftFormatter()
+        self._cursor_clear_formatter = ClearLineFormatter()
+
+    def stream_progress(self, logs: docker.APIClient.logs):
+        """
+        Stream progress from docker push logs and move the cursor based on the log id.
+        :param logs: generator from docker_clent.APIClient.logs
+        """
+        ids: Dict[str, int] = dict()
+        for log in logs:
+            _id = log.get("id", "")
+            status = log.get("status", "")
+            stream = log.get("stream", "")
+            progress = log.get("progress", "")
+            error = log.get("error", "")
+            change_cursor_count = 0
+            if _id:
+                try:
+                    curr_log_line_id = ids[_id]
+                    change_cursor_count = len(ids) - curr_log_line_id
+                    self._stream.write(
+                        self._cursor_up_formatter.cursor_format(change_cursor_count)
+                        + self._cursor_left_formatter.cursor_format(),
+                        encode=True,
+                    )
+                except KeyError:
```

Comment:
I assume this is for catching missing key in the statement of `curr_log_line_id = ids[_id]`.
If so, let's use `if _id in ids:` instead of using exceptions to control flow.

Reply:
Good call.